### PR TITLE
ikev1: add ikev1 parser - v9

### PR DESCRIPTION
**Pull request by fhonza**
Previous PR: #5439 

Describe changes:
- fixed and updated tests
- documentation update
- support eve multi-threaded output

suricata-verify-pr: 338

### CODE REVIEW

**Comment by fhonza** [comment]
will fix the cocci test...

**Comment by catenacyber** [comment]
Travis is failing because it does not use https://github.com/OISF/suricata-verify/pull/338

File: src/detect-engine-uint.c [hunk 1]
```diff
@@ -137,9 +147,17 @@ DetectU32Data *DetectU32Parse (const char *u32str)
                     return NULL;
 
                 if (arg2[0] == '<') {
-                    u32da.mode = DETECT_UINT_LT;
+                    if (arg2[1] == '=') {
```
Comment (catenacyber) [inline src/detect-engine-uint.c:150]: Thanks for these.
If you wish, could you remove the C implementation and use the Rust one from
https://github.com/OISF/suricata/blob/master/rust/src/http2/parser.rs#L757

File: src/app-layer-register.c [hunk 2]
```diff
@@ -189,6 +189,13 @@ int AppLayerRegisterParser(const struct AppLayerParser *p, AppProto alproto)
     return 0;
 }
 
+int AppLayerRegisterParserAlias(const char *proto_name, const char *proto_alias)
+{
+    AppLayerProtoDetectRegisterAlias(proto_name, proto_alias);
```
Comment (catenacyber) [inline src/app-layer-register.c:194]: Why does it make sense to have 2 functions ? `AppLayerRegisterParserAlias ` and `AppLayerProtoDetectRegisterAlias `

File: src/app-layer-detect-proto.c [hunk 3]
```diff
@@ -2041,6 +2093,15 @@ AppProto AppLayerProtoDetectGetProtoByName(const char *alproto_name)
 {
     SCEnter();
 
+    AppLayerProtoDetectAliases *cur_alias = alpda_ctx;
+    while (cur_alias != NULL) {
```
Comment (catenacyber) [inline src/app-layer-detect-proto.c:2101]: Why did you use a linked list ?

File: src/tests/fuzz/fuzz_sigpcap.c [hunk 4]
```diff
@@ -88,7 +88,7 @@ outputs:\n\
         - nfs\n\
         - smb\n\
         - tftp\n\
-        - ikev2\n\
+	- ike\n\
```
Comment (catenacyber) [inline src/tests/fuzz/fuzz_sigpcap.c]: Indent looks wrong

File: src/util-profiling.c [hunk 5]
```diff
@@ -1319,6 +1318,7 @@ const char * PacketProfileLoggertIdToString(LoggerId id)
         CASE_CODE (LOGGER_JSON_TEMPLATE_RUST);
         CASE_CODE (LOGGER_JSON_RFB);
         CASE_CODE (LOGGER_JSON_MQTT);
+        CASE_CODE(LOGGER_JSON_IKE);
```
Comment (catenacyber) [inline src/util-profiling.c:1321]: nit space

File: suricata.yaml.in [hunk 6]
```diff
@@ -714,6 +715,8 @@ pcap-file:
 # "detection-only" enables protocol detection only (parser disabled).
 app-layer:
   protocols:
+    ike:
```
Comment (catenacyber) [inline suricata.yaml.in:718]: Why did you move this up ?

File: src/output-json-ike.c [hunk 7]
```diff
@@ -0,0 +1,189 @@
+/* Copyright (C) 2020 Open Information Security Foundation
```
Comment (catenacyber) [inline src/output-json-ike.c:1]: Could you have first done `git mv output-json-ikev2.c output-json-ike.c `

Comment (catenacyber) [reply src/output-json-ike.c:1]: By the way, is there a set of bash commands like `sed` that you ran to get this commit ?

**Review by catenacyber — CHANGES_REQUESTED** [review]
This is looking good with CI happy :-)

There are still some questions back and forth to do
I did look into last commit yet

File: doc/userguide/output/eve/eve-json-format.rst [hunk 8]
```diff
@@ -1804,3 +1804,111 @@ Example of HTTP2 logging, of a request and response:
       ]
     }
   }
+
+Event type: IKE
+---------------
+
+The following fields and example are for IKEv1, IKEv2 has a different output format. IKEv1 is denoted by the field version_major = 1.
```
Comment (catenacyber) [inline doc/userguide/output/eve/eve-json-format.rst:1811]: You can say something like

You can distinguish between IKEv1 and IKEv2 using the version_major field (which equals either 1 or 2)

File: doc/userguide/rules/intro.rst [hunk 9]
```diff
@@ -89,7 +89,7 @@ you can pick from. These are:
 * dnp3 (disabled by default)
 * enip (disabled by default)
 * nfs
-* ikev2
+* ike
```
Comment (catenacyber) [inline doc/userguide/rules/intro.rst:92]: Should have this been in the previous commit ?

File: rust/src/ike/mod.rs [hunk 10]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2017-2018 Open Information Security Foundation
+/* Copyright (C) 2020 Open Information Security Foundation
```
Comment (catenacyber) [inline rust/src/ike/mod.rs:1]: Should now be 2017-2020

File: rust/src/ike/mod.rs [hunk 11]
```diff
@@ -14,11 +14,11 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
  * 02110-1301, USA.
  */
-
-// written by Pierre Chifflier  <chifflier@wzdftpd.net>
```
Comment (catenacyber) [inline rust/src/ike/mod.rs:18]: Why did you remove Pierre ?

File: rust/src/lib.rs [hunk 12]
```diff
@@ -67,7 +67,6 @@ pub mod smb;
 pub mod krb;
 pub mod dcerpc;
 
-pub mod ike;
```
Comment (catenacyber) [inline rust/src/lib.rs]: Why did you move this ?

File: src/app-layer-detect-proto.c [hunk 13]
```diff
@@ -179,7 +179,7 @@ struct AppLayerProtoDetectThreadCtx_ {
 
 /* The global app layer proto detection context. */
 static AppLayerProtoDetectCtx alpd_ctx;
-static AppLayerProtoDetectAliases *alpda_ctx;
+static AppLayerProtoDetectAliases *alpda_ctx = NULL;
```
Comment (catenacyber) [inline src/app-layer-detect-proto.c:182]: Should have this been in a previous commit ?

File: src/detect-engine-uint.c (hunk 1 above)
Comment (fhonza) [reply src/detect-engine-uint.c:150]: Hmm ok, a few questions:

Then i would call this rust-function from C in the *Setup()-function of the keyword?
This looks pretty generic, will this stay in the http2/parser?
Will detect-engine-uint.h/.c be deprecated in the future?

File: src/app-layer-register.c (hunk 2 above)
Comment (fhonza) [reply src/app-layer-register.c:194]: Thought that registering the aliases is semantically very similar to registering the protocols. Just thought that it's "the way" after seeing how AppLayerRegisterProtocolDetection() and AppLayerProtoDetectRegisterProtocol() are doing it.

That's why i placed alpda_ctx next to alpd_ctx in app-layer-detect-proto.c.

File: src/app-layer-detect-proto.c (hunk 3 above)
Comment (fhonza) [reply src/app-layer-detect-proto.c:2101]: My intention was to add a generic functionality that parsers can register arbitrary aliases, the linked list pattern is used often in suricata and i don't know before how many aliases will be registered....

File: suricata.yaml.in (hunk 6 above)
Comment (fhonza) [reply suricata.yaml.in:718]: don't know, i'll check

File: src/output-json-ike.c (hunk 7 above)
Comment (fhonza) [reply src/output-json-ike.c:1]: Normally i'm resetting my commits and use meld and a second suricata-folder to create the different commits, ensuring that they compile. But it's pretty painful with the amount of changes....
Do you have a better approach to that?

Did you mean that? :)

File: rust/src/ike/mod.rs (hunk 11 above)
Comment (fhonza) [reply rust/src/ike/mod.rs:18]: Not intentionally, perhaps happened when creating the commits....

File: src/output-json-ike.c (hunk 7 above)
Comment (catenacyber) [reply src/output-json-ike.c:1]: Well, I do not know if my approach would work in this case.
But I would have done 
one commit with `git mv "*ikev2*" "*ike*"` (and looking for the file names with git grep)
one commit like `find . -name "*.c" | sed -i -e s/ikev2/ike/g` (for the different files, and CamelCase and so on)

I do not have the exact syntax, but I would have tried something automated

File: src/app-layer-detect-proto.c (hunk 3 above)
Comment (catenacyber) [reply src/app-layer-detect-proto.c:2101]: An array would have seemed simpler to me, but that's fine I guess

File: src/app-layer-register.c (hunk 2 above)
Comment (catenacyber) [reply src/app-layer-register.c:194]: ok thanks

File: src/detect-engine-uint.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-engine-uint.c:150]: First, do not feel forced to do this...

If you want it, then 

> i would call this rust-function from C in the *Setup()-function of the keyword?

yes, and from the match function as well I guess

> This looks pretty generic, will this stay in the http2/parser?

No, if you can move it to somewhere generic, I would appreciate it

> Will detect-engine-uint.h/.c be deprecated in the future?

Yes, replace it with the rust version

So, do as you wish

File: src/output-json-ike.c (hunk 7 above)
Comment (fhonza) [reply src/output-json-ike.c:1]: thx, i'll give it a try

File: src/detect-engine-uint.c (hunk 1 above)
Comment (fhonza) [reply src/detect-engine-uint.c:150]: hmmm i would stick to the implemented solution, sry feels a bit more mature :)

Currently I've much other stuff on the table hope to get the v10 done soon....
